cmd/console-proxy/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! ## `humility console-proxy`
+//!
+//! Act as a proxy for the host serial console when it is jumpered to the SP.
+
+use std::path::PathBuf;
+
+use clap::{Command as ClapCommand, CommandFactory, Parser};
+
+use humility_cmd::{Archive, Attach, Command, Validate};
+
+#[cfg(not(windows))]
+mod posix;
+
+#[cfg(not(windows))]
+use posix::console_proxy;
+
+#[cfg(windows)]
+fn console_proxy(
+    _context: &mut humility::ExecutionContext,
+) -> anyhow::Result<()> {
+    anyhow::bail!("the console-proxy subcommand is not available on Windows")
+}
+
+#[derive(Parser, Debug)]
+#[clap(name = "console-proxy", about = env!("CARGO_PKG_DESCRIPTION"))]
+struct UartConsoleArgs {
+    #[clap(
+        long,
+        short = 'T',
+        default_value_t = 5000,
+        value_name = "hiffy_timeout_ms"
+    )]
+    hiffy_timeout: u32,
+
+    #[clap(
+        long,
+        short,
+        default_value_t = 1000,
+        help = "frequency of polling the SP for new data",
+        value_name = "poll_interval_ms"
+    )]
+    poll_interval: u32,
+
+    #[clap(subcommand)]
+    cmd: UartConsoleCommand,
+}
+
+#[derive(Parser, Debug)]
+enum UartConsoleCommand {
+    /// Attach Humility as the host console uart client via the
+    /// `control_plane_agent` task.
+    ///
+    /// By default, puts the current terminal into raw mode (use `--no-raw` to
+    /// disable this). In raw mode, exit via Ctrl-A Ctrl-X. Ctrl-A Ctrl-A will
+    /// send a Ctrl-A; no other prefixed commands are recognized.
+    ///
+    /// When Humility attaches as the uart client, it disables the SP's default
+    /// behavior of constantly reading and discarding data from the console
+    /// uart. If the attach process ends cleanly, it will detach and restore the
+    /// client setting to MGS (and restore the default behavior). To check
+    /// whether this happened, use the `client` subcommand. To forcibly set the
+    /// client back to MGS (e.g., if `attach` did not exit cleanly), use the
+    /// `detach` subcommand.
+    Attach {
+        #[clap(
+            long = "--no-raw",
+            help = "do not put terminal in raw mode",
+            action = clap::ArgAction::SetFalse,
+        )]
+        raw: bool,
+
+        #[clap(
+            long,
+            help = "Specifies the input character map (i.e., special characters to be replaced when reading from the serial port). See picocom's manpage."
+        )]
+        imap: Option<String>,
+
+        #[clap(
```

Comment:
I always find myself having to look up how to invoke `picocom` (`--imap lfcrlf --omap crlf,delbs`).  Should we make those the default?

Reply:
If I make those the defaults, how does one run this with no remappings enabled? (I'll admit it doesn't seem likely one would want to, but...)

Reply:
Would `--imap ""` disable them?  Or a separate `--no-imap`?

Reply:
Ah yeah `--imap ""` would work. Kinda weird, but I don't actually expect anyone to need to do that. I'll change this.